### PR TITLE
Limit Concurrent Session to 1 per oauth client #2733

### PR DESCRIPTION
**Pull request by tung2744**
ref #2733 #2756 #2757 #2758 #2759

Changes:
  - [x] Add x_max_concurrent_session to client config
  - [x] Revoke session of the same client when max concurrent session is set
  - [x] Screen to warn the end-user that other sessions will be terminated
  - [x] Update portal to configure x_max_concurrent_session
Screenshot 2023-01-05 at 7 36 54 PM

https://user-images.githubusercontent.com/34942210/210772473-f9a44732-b236-4c3e-a4b9-1e2a23913932.mov

### CODE REVIEW

File: pkg/lib/interaction/intents/authenticate.go [hunk 1]
```diff
@@ -357,6 +357,10 @@ func (i *IntentAuthenticate) DeriveEdgesForNode(graph *interaction.Graph, node i
 			panic(fmt.Errorf("interaction: unexpected authentication stage: %v", node.Stage))
 		}
 	case *nodes.NodePromptCreatePasskeyEnd:
+		return []interaction.Edge{
+			&nodes.EdgeConfirmTerminateOtherSessionsBegin{},
+		}, nil
```
Comment (carmenlau) [inline pkg/lib/interaction/intents/authenticate.go:363]: Currently, when I try to promote an anonymous user in the mobile app, I will see this warning screen. Since a refresh token was issued when the user is an anonymous user. Maybe we should only show this when login?

**Comment by tung2744** [comment]
TODO:

- [ ] Block using sso login with concurrent session config in authorize endpoint
- [ ] clear steps in session
- [ ] redirect user back to select account

File: pkg/lib/oauth/handler/handler_token.go [hunk 2]
```diff
@@ -701,6 +709,22 @@ func (h *TokenHandler) handleIDToken(
...
+	userID string) error {
+	offlineGrants, err := h.OfflineGrants.ListOfflineGrants(userID)
+	if err != nil {
+		return err
+	}
+	for _, offlineGrant := range offlineGrants {
+		if offlineGrant.ClientID != client.ClientID {
+			continue
+		}
+		h.SessionManager.RevokeWithEvent(offlineGrant, false, false)
```
[the remark this answers is not in the document]
Comment (carmenlau) [reply pkg/lib/oauth/handler/handler_token.go]: Discussion conclusion, having `x_max_concurrent_session` in the client + sso enabled is rare. We will block it from the authz endpoint.

**Comment by tung2744** [comment]
@carmenlau Updated, thx!

File: pkg/lib/oauth/handler/handler_token.go [hunk 3]
```diff
@@ -789,14 +808,7 @@ func (h *TokenHandler) issueTokensForAuthorizationCode(
...
 	if issueRefreshToken {
-		// First revoke existing refresh tokens if MaxConcurrentSession == 1
-		if client.MaxConcurrentSession == 1 {
-			err := h.revokeClientOfflineGrants(client, code.AuthenticationInfo.UserID)
-			if err != nil {
-				return nil, err
-			}
-		}
-		offlineGrant, err := h.TokenService.IssueOfflineGrant(client, opts, resp)
+		offlineGrant, err := h.issueOfflineGrant(client, code, resp, opts)
```
Comment (carmenlau) [inline pkg/lib/oauth/handler/handler_token.go:811]: In the `TokenHandler`, we have multiple places that call `h.TokenService.IssueOfflineGrant` to issue the refresh token. Should we also update those places to use `h.issueOfflineGrant` instead? (e.g. `handleBiometricAuthenticate`, to restrict biometric login concurrent sessions ...etc.) If yes, please check the functionality after updating

File: pkg/auth/webapp/service2.go [hunk 4]
```diff
@@ -332,16 +332,7 @@ func (s *Service2) doPost(
...
-	switch graph.CurrentNode().(type) {
-	case *nodes.NodeCancelInteraction:
-		isCancelled = true
-	}
-	if isCancelled {
-		err = nil
-	}
-
-	err = s.afterPost(result, session, graph, err, isCancelled, isFinished, isNewGraph)
+	err = s.afterPost(result, session, graph, err, isFinished, isNewGraph)
```
Comment (carmenlau) [inline pkg/auth/webapp/service2.go:335]: It is okay for now. But if we want to revert some unwanted changes, it would be better to do it in a separate commit or even update the commit through rebase. So that we can have a clear commit history.

**Review by carmenlau — COMMENTED** [review]
Please rebase

File: pkg/lib/oauth/handler/handler_token.go (hunk 3 above)
Comment (tung2744) [reply pkg/lib/oauth/handler/handler_token.go:811]: I guess yes, let me check it

**Comment by tung2744** [comment]
@carmenlau Updated. Also checked biometric login is working. Thx!

**Comment by carmenlau** [comment]
Please fix the build, thanks @tung2744!!

**Comment by tung2744** [comment]
@carmenlau Fixed and rebased again, thx!
